xtm/xtm_api.h

```diff
@@ -0,0 +1,179 @@
+#pragma once
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <stdlib.h>
+#include <stdbool.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/** \cond public */
+
+/**
+ * Opaque struct, that represent unidirectional, single-writer-single-reader
+ * queue implementation with event loop integration ability.
+ */
+struct xtm_queue;
+
+/**
+ * Typedef for often used function pointer, used in push_fun/invoke_funs_all
+ * xtm used pattern.
+ */
+typedef void (*xtm_queue_fun_t)(void*);
+
+/**
+ * Create instance of struct xtm_queue.
+ * @param[in] size                           - queue size, must be power of two
+ * @param[in] is_need_to_close_readfd        - flag indicates, that we need to close
+ *                                             all queue read file descriptors manually.
+ * @param[in] is_producer_need_notifications - flag indicates, that producer,
+ *                                             wants to get notification about free
+ *                                             place in queue.
+ * @retval    pointer to new xtm_queue or NULL in case of error.
+ */
+struct xtm_queue *
+xtm_queue_new(unsigned size, bool is_need_to_close_readfd,
+	      bool is_producer_need_notifications);
```

Comment:
Yes, good idea! I add xtm_queue_attr_set function too. If there are any comments on the implementation, I will correct them!

Reply:
Maybe "unsigned flags" would suffice?

Reply:
What if you will need to pass not only the flags? What if some other event loop implementation will need us to take its own descriptors instead of ours? What if eventually we would allow to customize the queue type (linked list, array like now, multiple producers, etc) (although then it would be better to have separate queue structs)? `unsigned` type is not extendible, it is a dead end for the highly mutable API like we have here.

Reply:
> What if some other event loop implementation will need us to take its own descriptors instead of ours?

A lot of things are possible but highly unlikely. Overkill.

Reply:
The fact that we can't close the descriptors manually was also an unlikely event, and yet it happened. I don't want to have `xtm_queue_new_ex` in the future which will end up using `attr` anyway. But ok, the worst what can happen we will create an `ex` function. I am fine with the `flags` bitmask argument.